idaes/models_extra/power_generation/unit_models/soc_submodels.py

```diff
@@ -0,0 +1,3997 @@
+##############################################################################
+# Institute for the Design of Advanced Energy Systems Process Systems
+# Engineering Framework (IDAES PSE Framework) Copyright (c) 2018-2020, by the
+# software owners: The Regents of the University of California, through
+# Lawrence Berkeley National Laboratory,  National Technology & Engineering
+# Solutions of Sandia, LLC, Carnegie Mellon University, West Virginia
+# University Research Corporation, et al. All rights reserved.
+#
+# Please see the files COPYRIGHT.txt and LICENSE.txt for full copyright and
+# license information, respectively. Both files are also available online
+# at the URL "https://github.com/IDAES/idaes-pse".
+##############################################################################
+
+import copy
+import enum
+
+from pyomo.common.config import ConfigBlock, ConfigValue, In
+from pyomo.dae import DerivativeVar
+import pyomo.environ as pyo
+from pyomo.network import Port
+
+
+from idaes.core import declare_process_block_class, UnitModelBlockData, useDefault
+from idaes.models.unit_models.heat_exchanger import HeatExchangerFlowPattern
+import idaes.core.util.scaling as iscale
+from idaes.core.util.exceptions import ConfigurationError, InitializationError
+from idaes.core.util.math import safe_log
+from idaes.core.util import get_solver
+import idaes.core.util.model_statistics as mstat
+
+from idaes.core.util.misc import VarLikeExpression
+import idaes.logger as idaeslog
+
+
+
+_constR = 8.3145 * pyo.units.J / pyo.units.mol / pyo.units.K  # or Pa*m3/K/mol
+_constF = 96485 * pyo.units.coulomb / pyo.units.mol
+_scaling_factor_pressure = 1E-4
+_scaling_factor_temperature_thermo = 1E-2
+_scaling_factor_temperature_channel = 1E2
+_scaling_factor_temperature_conduction = 1E2
+_scaling_factor_concentration = 1
+_scaling_factor_Dconcentration = 10
+_scaling_factor_energy = 1e-3
+_safe_log_eps = 1e-9
+_safe_sqrt_eps = 1e-9
+_Tmax = None#900+273.15+300
+
+class CV_Bound(enum.Enum):
+    EXTRAPOLATE = 1
+    NODE_VALUE = 2
+
+
+class CV_Interpolation(enum.Enum):
+    UDS = 1  # Upwind difference scheme, exit face same as node center
+    CDS = 2  # Linear interpolation from upstream and downstream node centers
+    QUICK = 3  # Quadratic upwind interpolation, quadratic from two upwind
+    # centers and one downwind
+
+
+def _set_default_factor(c, s):
```

Comment:
These really look like they should be general purpose functions in the scaling toolbox.

Reply:
My preferred solution is to add an `overwrite=True/False` option to `set_scaling_factor` to handle the "if None` part, then make the "iterating over the index to set individual scaling factors" aspect default behavior of `set_scaling_factor`.  Indexed variables shouldn't have scaling factors distinct from their indexed parts.  `_set_and_get_scaling_factor` might still be a function then, but the other three are eliminated.

Reply:
I'm pretty sure we can add the overwrite option, and default to whatever it does now without hurting anything.

Reply:
Yeah, that change is pretty small. It's the other two that are more impactful. I maintain that calling `get_scaling_factor` on an indexed variable without specifying the indices should return an error: otherwise the user could specify scaling factors at each index yet calling `get_scaling_factor` on the variable for an unspecified index wouldn't return any factor.